subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/ProcessFixture.groovy

```diff
@@ -65,6 +65,27 @@ class ProcessFixture {
         return bash("ps -o pid,ppid,args -p ${pids.join(' -p ')}").split("\\n")
     }
 
+    void waitForFinish() {
+        if (pid == null) {
+            throw new RuntimeException("Unable to wait for process to finish because provided pid is null!")
+        }
+        if (OperatingSystem.current().unix) {
+            bash("wait $pid")
```

Comment:
A few error conditions....
- I think there's a race between the process exiting from the kill and this wait starting. What happens if the process exits quickly enough that the wait hasn't happened yet?
- What happens if the process never exits?

Reply:
> I think there's a race between the process exiting from the kill and this wait starting. What happens if the process exits quickly enough that the wait hasn't happened yet?

I thought about this, and I didn't think this would be a problem since `wait` on *nix handles this, and the windows implementation here handles it, but then when you asked about this, I realized when we execute the `wait` with `bash` we're putting it into a different process hierarchy in which case `wait` will complain since the pid is not a child and return a non-zero exit code.  So I changed this to a polling ps command, similar to how the windows implementation works.

> What happens if the process never exits?

I don't think there's much danger of this since we are sending uninterruptable kill signals (or rather, if we do have this problem, there are much bigger issues) but with the use of `ConcurrentTestUtil.poll()` it will now timeout after 10s if the process somehow doesn't die.